lib/Backends/OpenCL/OpenCL.cpp

```diff
@@ -1452,141 +1441,181 @@ uint64_t OpenCLFunction::copyValueFromDevice(const Value *v, void *buf) {
   return copiedBytes;
 }
 
-uint64_t OpenCLFunction::copyMutableWeightsToDevice() {
-  uint64_t copiedBytes = 0;
-  for (auto it : tensors_) {
-    if (!externalTensors_.count(it.first)) {
-      continue;
-    }
-    if (auto *W = dyn_cast<WeightVar>(it.first)) {
-      if (W->getMutability() == WeightVar::MutabilityKind::Constant)
-        continue;
-    }
-    copiedBytes += copyValueToDevice(it.first);
+void OpenCLFunction::copyConstantsToDevice() {
+  deviceBuffer_ = allocDeviceBuffer(bundle_.constantWeightVarsMemSize +
```

Comment:
You allocate all of the required memory in copyConstantsToDevice, including constants, mutable weights and activations. It seems wrong. The buffer should be allocated before this function is called.

Reply:
This will be changed as part of the ongoing work for #1904. A new api is called out there which has more appropriate names and division of work.

Reply:
Can you still add the comment here about the plans to move this allocation out of this function in the future?